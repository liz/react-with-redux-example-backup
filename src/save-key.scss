html, body, #root, *[class*='container__Shell'] {
	height: 100%;
}