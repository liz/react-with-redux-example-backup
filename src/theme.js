// eslint-disable-next-line import/no-webpack-loader-syntax
import variables from 'sass-extract-loader?{"plugins" :["sass-extract-js"]}!./variables.scss';

export default variables;