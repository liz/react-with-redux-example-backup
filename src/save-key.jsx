import React, { useState, useEffect } from 'react';
import styled from 'styled-components/macro';
import { connect } from 'react-redux'

import { saveKey } from './actions';

import theme from './theme';
import mediaQueries from './media-queries';

import { Container } from './components/container';
import { Row } from './components/row';
import { FormInput } from './components/form-input';
import { Button } from './components/button';

import './save-key.scss';

const Col = styled.div`
    width: 100%;
    padding-left: ${theme.gutter};
    padding-right: ${theme.gutter};
    text-align: center;
`;
Col.displayName = 'Col';


const OuterCol = styled(Col)`
    height: 100%;
    margin: auto;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;

    form {
        width: 100%;
        display: block;
        max-width: ${mediaQueries.min.iphone6};
    }
`;
OuterCol.displayName = 'OuterCol';

export const SaveKey = (props) => {
    const [fieldValue, setFieldValue] = useState('');
    const [buttonDisabled, setButtonDisabled] = useState(true);
    const [fieldError, setFieldError] = useState(...props.fieldError);

    useEffect(() => {
        setFieldError(props.fieldError);

        if (fieldValue) {
            setButtonDisabled(false)
        } else {
            setButtonDisabled(true)
        }
    }, [props.fieldError, fieldValue, fieldError]);

    const onSubmit = (event) => {
        event.preventDefault();
        if (fieldValue) {
            props.dispatch(saveKey(fieldValue))
            setButtonDisabled(true)
        } else {
            setFieldError("Please enter an API Key")
        }
    }; 

    const onFieldChange = (event) => {
        setFieldValue(event.target.value);
    }; 

    return (
        <Container>
            <Row height="100%">
                <OuterCol>
                    <h1>Github Repo Issues</h1>
                    <form
                        onSubmit={(event) => onSubmit(event)}
                    >
                        <fieldset>
                            <Row>
                                <Col>
                                    <FormInput
                                        value={fieldValue}
                                        fieldChange={event => onFieldChange(event)} 
                                        placeHolder="Github API Key"
                                        fieldError={fieldError}
                                        fieldId="save-key"
                                        fieldLabel="Please submit your Github API Key to see issues for your repos"
                                    />
                                </Col>
                            </Row>
                            <Row>
                                <Col>
                                    <Button 
                                        type="submit"
                                        buttonText="Submit"
                                        disabled={buttonDisabled}
                                    />
                                </Col>
                            </Row>
                        </fieldset>
                    </form>
                </OuterCol>
            </Row>
        </Container>
    );
}

SaveKey.defaultProps = {
    fieldError: '',
};

export default connect()(SaveKey)