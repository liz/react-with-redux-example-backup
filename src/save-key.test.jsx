import React from 'react';
import { shallow, mount } from 'enzyme';
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import { act } from 'react-dom/test-utils';

import { SaveKey } from './save-key';
import theme from './theme';

const mockStore = configureMockStore();
const store = mockStore({});

describe('SaveKey', () => {
	let wrapper;

	beforeEach(() => {
	    wrapper = mount(
			<Provider store={store}>
				<SaveKey />
			</Provider>
		);
    });

	afterEach(() => {
		jest.resetAllMocks();
    });

	describe('Renders', () => {
		it('should match the snapshot', () => {
			expect(wrapper.html()).toMatchSnapshot();
	  	});

	  	it('renders FormInput with expected props', () => {
			expect(wrapper.find('FormInput')).toHaveLength(1);
			expect(wrapper.find('FormInput').props()).toEqual({
				value: '',
				fieldChange: expect.any(Function),
				placeHolder: 'Github API Key',
				fieldId: 'save-key',
				fieldLabel: 'Please submit your Github API Key to see issues for your repos',
				fieldType: 'text',
				fieldError:  '',
				disabled: false,
				required: false,
				rows: 2,
				bottomSpacing: '1rem'
			});
	  	});

	  	 it('renders submit Button with expected props', () => {
	  	 	expect(wrapper.find('Button')).toHaveLength(1);
			expect(wrapper.find('Button').props()).toEqual({
				type: 'submit',
				buttonText: 'Submit',
				disabled: true,
				color: '#d62027',
				colorAlt: null,
				iconOnRight: false,
				className: 'btn',
				hidden: false,
				minWidth: '175px'
			});
	  	});
	});

	describe('Updates fieldValue state', () => {
		it('disables submit button on load when form input field is empty', () => {
			expect(wrapper.find('#save-key').prop('value')).toEqual('');
			expect(wrapper.find('Button').prop('disabled')).toBe(true);
		});

		it('enables submit button when form input field has text', () => {
			expect(wrapper.find('#save-key').props().value).toEqual('');

			act(() => { 
				wrapper.find('#save-key').instance().value = '1234567900';
				wrapper.find('#save-key').simulate('change');
			});

			wrapper.update();

			expect(wrapper.find('#save-key').props().value).toEqual('1234567900');
			expect(wrapper.find('Button').prop('disabled')).toBe(false);
		});

		it('calls the dispatch function with expected params and disables the submit button on form submit', () => {
			const dispatch = jest.fn();

			wrapper = mount(
				<Provider store={store}>
					<SaveKey dispatch={dispatch} />
				</Provider>
			);

			expect(wrapper.find('#save-key').props().value).toEqual('');
			expect(wrapper.find('Button').prop('disabled')).toBe(true);

			act(() => { 
				wrapper.find('#save-key').instance().value = '1234567900';
				wrapper.find('#save-key').simulate('change');
			});

			wrapper.update();

			expect(wrapper.find('#save-key').props().value).toEqual('1234567900');
			expect(wrapper.find('Button').prop('disabled')).toBe(false);

			act(() => { 
				wrapper.find('form').simulate('submit');
			});

			wrapper.update();

			const expectedParams = {"key": "1234567900", "type": "SAVE_KEY"};

			expect(wrapper.find('Button').prop('disabled')).toBe(true);
			expect(dispatch).toHaveBeenCalledWith(expectedParams);
		});
	});

	describe('Updates fieldError state', () => {
		it('passes fieldError prop if supplied to the FormInputError component via the fieldError state', () => {
			wrapper = mount(
				<Provider store={store}>
					<SaveKey fieldError="Github does not recognize this API Key, please try a different API Key." />
				</Provider>
			);

			expect(wrapper.find('FormInput').find('FormInputError').props().fieldError).toEqual(wrapper.find('SaveKey').props().fieldError);
		});
	});
});