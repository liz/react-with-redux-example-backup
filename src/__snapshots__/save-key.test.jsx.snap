// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaveKey Renders should match the snapshot 1`] = `"<div class=\\"container__Shell-sc-1efg81n-0 gepMUD\\"><div height=\\"100%\\" class=\\"row__RowContainer-g7667b-0 hPpvDI\\"><div class=\\"save-key__Col-sc-1bv548z-0 save-key__OuterCol-sc-1bv548z-1 cbeJBp\\"><h1>Github Repo Issues</h1><form><fieldset><div class=\\"row__RowContainer-g7667b-0 fhhpBZ\\"><div class=\\"save-key__Col-sc-1bv548z-0 ckioLT\\"><div class=\\"form-input__FieldContainer-sc-18od1hf-0 dOvUPP\\"><label for=\\"save-key\\">Please submit your Github API Key to see issues for your repos</label><input id=\\"save-key\\" type=\\"text\\" name=\\"save-key\\" placeholder=\\"Github API Key\\" rows=\\"2\\" value=\\"\\"></div></div></div><div class=\\"row__RowContainer-g7667b-0 fhhpBZ\\"><div class=\\"save-key__Col-sc-1bv548z-0 ckioLT\\"><button color=\\"#d62027\\" class=\\"button__ColoredButton-sc-1gd83n3-0 bPICVJ btn\\" title=\\"Submit\\" type=\\"submit\\" aria-label=\\"Submit\\" disabled=\\"\\"><span class=\\"button__Content-sc-1gd83n3-1 kdVsyS\\"><span class=\\"button__Text-sc-1gd83n3-3 ZwnKy\\">Submit</span></span></button></div></div></fieldset></form></div></div></div>"`;
